kernel/arch/dreamcast/include/dc/maple/keyboard.h

```diff
@@ -28,283 +40,647 @@ __BEGIN_DECLS
+        uint8_t unknown2    : 1;    /**< \brief Unknown LED 2 */
+        uint8_t kana        : 1;    /**< \brief Kana LED */
+        uint8_t power       : 1;    /**< \brief Power LED */
+        uint8_t shift       : 1;    /**< \brief Shift LED */
+    };
+    uint8_t raw;    /**< \brief Packed 8-bit unsigned integer of bitflags */
+} kbd_leds_t;
+
 /** @} */
 
-/** \defgroup   kbd_regions Region Codes
-    \brief                  Region codes for the Dreamcast keyboard
-    \ingroup                kbd
+/** \defgroup key_state_grp     Key States
+    \brief                      Types associated with key states
+    \ingroup                    kbd_status_grp
+
+    Key states are represented by the kbd_state_t union type. The state may be
+    accessed by:
+        1. Directly comparing key_state_t::value to a \ref key_state_value_t.
+        2. Directly using a convenience bit field.
+        3. Bitwise `AND` of key_state_t::raw with one of the
+           \ref key_state_flags.
 
-    This is the list of possible values for the "region" field in the
-    kbd_state_t structure.
     @{
 */
-#define KBD_REGION_JP       1           /**< \brief Japanese keyboard */
-#define KBD_REGION_US       2           /**< \brief US keyboard */
-#define KBD_REGION_UK       3           /**< \brief UK keyboard */
-#define KBD_REGION_DE       4           /**< \brief German keyboard */
-#define KBD_REGION_FR       5           /**< \brief French keyboard */
-#define KBD_REGION_IT       6           /**< \brief Italian keyboard (not supported yet) */
-#define KBD_REGION_ES       7           /**< \brief Spanish keyboard */
-/** @} */
 
-/** \defgroup   key_states  Key States
-    \brief                  States each key can be in.
-    \ingroup                kbd
+/** \defgroup   key_state_flags Flags
+    \brief                      Keyboard key state bit flags
 
-    These are the different 'states' each key can be in. They are stored in
-    kbd_state_t->matrix, and manipulated/checked by kbd_check_poll.
+    A key_state_t is a combination of two flags:
+        1. `KEY_STATE_IS_DOWN`: whether the key is currently pressed
+                                this frame.
+        2. `KEY_STATE_WAS_DOWN`: whether the key was previously pressed
+                                 last frame.
+
+    Between these two flags, you can know whether a key state transition
+    event has occurred (when the two flags have different values).
+
+    \sa key_state_t::raw, key_state_value_t
 
-    none-> pressed or none
-    was pressed-> pressed or none
-    pressed-> was_pressed
     @{
 */
-#define KEY_STATE_NONE        0
-#define KEY_STATE_WAS_PRESSED 1
-#define KEY_STATE_PRESSED     2
+#define KEY_STATE_IS_DOWN    (1 << 0) /**< \brief If key is currenty down */
+#define KEY_STATE_WAS_DOWN   (1 << 1) /**< \brief If key was previously down */
+/** \brief Mask of all key state flags */
+#define KEY_STATE_MASK      (KEY_STATE_IS_DOWN | KEY_STATE_WAS_DOWN)
```

Comment:
Thanks, this is actually my second pass at a system like this, because I had one going in my original engine, then I took what I learned there and consulted what Simulant and Kazade were doing when constructing this. lol.